File: src/main/java/me/wolfii/implementations/day10/Direction.java
package me.wolfii.implementations.day10;

import me.wolfii.implementations.common.Vec2;

public enum Direction {
    NORTH(0, -1),
    SOUTH(0, 1),
    EAST(1, 0),
    WEST(-1, 0),
    NONE(0, 0),
    UNKNOWN(0, 0);
    final Vec2 vec2;

    Direction(int x, int y) {
        vec2 = new Vec2(x, y);
    }

    public static Direction of(Vec2 vector) {
        if (vector.equals(new Vec2(0, -1))) {
            return NORTH;
        } else if (vector.equals(new Vec2(0, 1))) {
            return SOUTH;
        } else if (vector.equals(new Vec2(1, 0))) {
            return EAST;
        } else if (vector.equals(new Vec2(-1, 0))) {
            return WEST;
        }
        return UNKNOWN;
    }

    public Direction turnRight() {
        return switch (this) {
            case NORTH -> EAST;
            case EAST -> SOUTH;
            case SOUTH -> WEST;
            case WEST -> NORTH;
            default -> UNKNOWN;
        };
    }

    public Direction turnLeft() {
        return switch (this) {
            case NORTH -> WEST;
            case WEST -> SOUTH;
            case SOUTH -> EAST;
            case EAST -> NORTH;
            default -> UNKNOWN;
        };
    }

    public Direction inverted() {
        return switch (this) {
            case NORTH -> SOUTH;
            case SOUTH -> NORTH;
            case EAST -> WEST;
            case WEST -> EAST;
            default -> this;
        };
    }
}

File: 2024/src/main/kotlin/common/Direction.kt
package me.wolfii.common

enum class Direction(val vec: Vec2I) {
    NORTH(Vec2I(0, -1)),
    EAST(Vec2I(1, 0)),
    SOUTH(Vec2I(0, 1)),
    WEST(Vec2I(-1, 0));

    fun turnRight() = when (this) {
        NORTH -> EAST
        EAST -> SOUTH
        SOUTH -> WEST
        WEST -> NORTH
    }

    fun turnLeft() = when (this) {
        NORTH -> WEST
        EAST -> NORTH
        SOUTH -> EAST
        WEST -> SOUTH
    }

    fun inverse() = when (this) {
        NORTH -> SOUTH
        SOUTH -> NORTH
        EAST -> WEST
        WEST -> EAST
    }
}